File: web/public/src/lib/issues.ts
import fs from "fs";
import path from "path";
import { getEditionInfo } from "./edition";

const ROOT = process.env.BRIEF_ROOT ?? path.join(process.cwd(), "..", "..");
const OUTPUT_DIR = path.join(ROOT, "output");
const PLACEHOLDER_PATH = path.join(ROOT, "content", "placeholder", "issue.json");

export interface PublicStory {
  title: string;
  url: string;
  source_name: string;
  category: string;
  summary: string;
  why_it_matters: string;
  read_time_minutes: number;
  apac_score: number;
}

export interface PublicIssue {
  date: string;
  edition_slug: string;
  intro: string;
  apac_ratio: number;
  is_sample: boolean;
  stories: PublicStory[];
}

export interface IssueSummary {
  date: string;
  story_count: number;
  is_sample: boolean;
  apac_ratio: number;
}

// Issue dates become filesystem path components — accept plain YYYY-MM-DD only.
const ISSUE_DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

function isValidIssueDate(value: string): boolean {
  return ISSUE_DATE_RE.test(value);
}

function loadJsonIssue(filePath: string): PublicIssue | null {
  if (!fs.existsSync(filePath)) return null;
  try {
    const data = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    if (typeof data.date !== "string") throw new Error("missing date");
    return {
      date: data.date,
      edition_slug: data.edition_slug,
      intro: data.intro ?? "",
      apac_ratio: Number(data.apac_ratio ?? 0),
      is_sample: Boolean(data.is_sample),
      stories: (data.stories ?? []).map((s: PublicStory) => ({
        title: s.title,
        url: s.url,
        source_name: s.source_name,
        category: s.category ?? "misc",
        summary: s.summary ?? "",
        why_it_matters: s.why_it_matters ?? "",
        read_time_minutes: Number(s.read_time_minutes ?? 3),
        apac_score: Number(s.apac_score ?? 0),
      })),
    };
  } catch (error) {
    console.warn(`Skipping malformed issue file ${filePath}:`, error);
    return null;
  }
}

function editionSlug(): string {
  return getEditionInfo().slug;
}

function listPublishedDates(slug: string): string[] {
  const editionDir = path.join(OUTPUT_DIR, slug);
  if (!fs.existsSync(editionDir)) return [];
  return fs
    .readdirSync(editionDir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .filter(isValidIssueDate)
    .filter((name) => fs.existsSync(path.join(editionDir, name, "issue.json")))
    .sort()
    .reverse();
}

export function loadPlaceholderIssue(): PublicIssue {
  const issue = loadJsonIssue(PLACEHOLDER_PATH);
  if (!issue) throw new Error(`Placeholder issue missing at ${PLACEHOLDER_PATH}`);
  return issue;
}

export function loadPublishedIssue(issueDate: string): PublicIssue | null {
  if (!isValidIssueDate(issueDate)) return null;
  const slug = editionSlug();
  return loadJsonIssue(path.join(OUTPUT_DIR, slug, issueDate, "issue.json"));
}

export function listPublicIssues(includeSample = true): IssueSummary[] {
  const slug = editionSlug();
  const summaries: IssueSummary[] = [];
  for (const issueDate of listPublishedDates(slug)) {
    const issue = loadPublishedIssue(issueDate);
    if (issue) {
      summaries.push({
        date: issue.date,
        story_count: issue.stories.length,
        is_sample: false,
        apac_ratio: issue.apac_ratio,
      });
    }
  }
  if (includeSample) {
    const sample = loadPlaceholderIssue();
    summaries.push({
      date: sample.date,
      story_count: sample.stories.length,
      is_sample: true,
      apac_ratio: sample.apac_ratio,
    });
  }
  return summaries;
}

export function getPublicIssue(issueDate: string): PublicIssue | null {
  if (issueDate === "sample") return loadPlaceholderIssue();
  const published = loadPublishedIssue(issueDate);
  if (published) return published;
  if (issueDate === loadPlaceholderIssue().date) return loadPlaceholderIssue();
  return null;
}

export function getFeaturedIssue(): PublicIssue {
  const dates = listPublishedDates(editionSlug());
  if (dates.length > 0) {
    const issue = loadPublishedIssue(dates[0]);
    if (issue) return issue;
  }
  return loadPlaceholderIssue();
}

File: brief/issues.py
from __future__ import annotations

import json
import logging
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from brief.entities import Issue
from brief.models import load_edition_config
from brief.paths import OUTPUT_DIR, ROOT

logger = logging.getLogger(__name__)

PLACEHOLDER_PATH = ROOT / "content" / "placeholder" / "issue.json"

ISSUE_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def is_valid_issue_date(value: str) -> bool:
    """Issue dates are path components — reject anything that is not
    a plain YYYY-MM-DD string before it reaches the filesystem."""
    return bool(ISSUE_DATE_RE.fullmatch(value))


@dataclass
class PublicStory:
    title: str
    url: str
    source_name: str
    category: str
    summary: str
    why_it_matters: str
    read_time_minutes: int
    apac_score: float


@dataclass
class PublicIssue:
    date: str
    edition_slug: str
    intro: str
    apac_ratio: float
    is_sample: bool
    stories: list[PublicStory]

    @property
    def story_count(self) -> int:
        return len(self.stories)

    @property
    def label(self) -> str:
        if self.is_sample:
            return "Sample issue"
        return self.date


@dataclass
class IssueSummary:
    date: str
    story_count: int
    is_sample: bool
    apac_ratio: float


def category_labels() -> dict[str, str]:
    edition = load_edition_config()
    return {item["slug"]: item["label"] for item in edition.get("categories", [])}


def edition_info() -> dict[str, Any]:
    return load_edition_config()["edition"]


def _story_from_dict(data: dict[str, Any]) -> PublicStory:
    return PublicStory(
        title=data["title"],
        url=data["url"],
        source_name=data["source_name"],
        category=data.get("category", "misc"),
        summary=data.get("summary", ""),
        why_it_matters=data.get("why_it_matters", ""),
        read_time_minutes=int(data.get("read_time_minutes", 3)),
        apac_score=float(data.get("apac_score", 0.0)),
    )


def issue_from_dict(data: dict[str, Any]) -> PublicIssue:
    return PublicIssue(
        date=data["date"],
        edition_slug=data.get("edition_slug", edition_info()["slug"]),
        intro=data.get("intro", ""),
        apac_ratio=float(data.get("apac_ratio", 0.0)),
        is_sample=bool(data.get("is_sample", False)),
        stories=[_story_from_dict(item) for item in data.get("stories", [])],
    )


def issue_to_dict(issue: Issue, apac_ratio: float, is_sample: bool = False) -> dict[str, Any]:
    return {
        "date": issue.date,
        "edition_slug": issue.edition_slug,
        "title": issue.title,
        "intro": issue.intro,
        "apac_ratio": apac_ratio,
        "is_sample": is_sample,
        "stories": [
            {
                "title": story.title,
                "url": story.url,
                "source_name": story.source_name,
                "category": story.category,
                "summary": story.summary,
                "why_it_matters": story.why_it_matters,
                "read_time_minutes": story.read_time_minutes,
                "apac_score": story.apac_score,
            }
            for story in issue.stories
        ],
    }


def load_json_issue(path: Path) -> PublicIssue | None:
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return issue_from_dict(data)
    except (json.JSONDecodeError, KeyError, TypeError, ValueError) as exc:
        logger.warning("Skipping malformed issue file %s: %s", path, exc)
        return None


def published_issue_dir(edition_slug: str, issue_date: str) -> Path:
    return OUTPUT_DIR / edition_slug / issue_date


def list_published_dates(edition_slug: str | None = None) -> list[str]:
    slug = edition_slug or edition_info()["slug"]
    edition_dir = OUTPUT_DIR / slug
    if not edition_dir.exists():
        return []
    dates = []
    for child in edition_dir.iterdir():
        if child.is_dir() and is_valid_issue_date(child.name) and (child / "issue.json").exists():
            dates.append(child.name)
    return sorted(dates, reverse=True)


def load_published_issue(issue_date: str, edition_slug: str | None = None) -> PublicIssue | None:
    if not is_valid_issue_date(issue_date):
        return None
    slug = edition_slug or edition_info()["slug"]
    return load_json_issue(published_issue_dir(slug, issue_date) / "issue.json")


def load_placeholder_issue() -> PublicIssue:
    issue = load_json_issue(PLACEHOLDER_PATH)
    if issue is None:
        raise FileNotFoundError(f"Placeholder issue missing at {PLACEHOLDER_PATH}")
    return issue


def list_public_issues(include_sample: bool = True) -> list[IssueSummary]:
    slug = edition_info()["slug"]
    summaries: list[IssueSummary] = []
    for issue_date in list_published_dates(slug):
        issue = load_published_issue(issue_date, slug)
        if issue:
            summaries.append(
                IssueSummary(
                    date=issue.date,
                    story_count=issue.story_count,
                    is_sample=False,
                    apac_ratio=issue.apac_ratio,
                )
            )
    if include_sample:
        sample = load_placeholder_issue()
        summaries.append(
            IssueSummary(
                date=sample.date,
                story_count=sample.story_count,
                is_sample=True,
                apac_ratio=sample.apac_ratio,
            )
        )
    return summaries


def get_public_issue(issue_date: str) -> PublicIssue | None:
    if issue_date == "sample":
        return load_placeholder_issue()
    published = load_published_issue(issue_date)
    if published:
        return published
    if issue_date == load_placeholder_issue().date:
        return load_placeholder_issue()
    return None


def get_featured_issue() -> PublicIssue:
    dates = list_published_dates()
    if dates:
        issue = load_published_issue(dates[0])
        if issue:
            return issue
    return load_placeholder_issue()
